### DISCUSSION TITLE
drop-in replacement for read-installed

### DISCUSSION THREAD

**Discussion by link89** [npm/feedback#539] [CLI]
The [read-installed](https://github.com/npm/read-installed) has been deprecated and its latest release was about 6 yrs ago.
Is there any drop-in replacement or successor for it? It is a common requirement to read installed packages and I believe there must be official implementation to do this. It will be appreciated if someone can shed me some light.

**Comment by MylesBorins** [comment]
pinging the @npm/cli team for comment

**Comment by ljharb** [comment]
I believe https://npmjs.com/@npmcli/arborist is the likely package to use at this point.

**Reply by isaacs** [reply]
It's far from a "drop-in" replacement, but yes, `@npmcli/arborist` is the successor and the thing we use now.

Once upon a time, `read-installed` was sort of a precursor to `read-package-tree`, which was less opinionated, had more features, and enabled better tree-walking and tree-munging operations.  The only place `read-installed` was still used was in `lib/rebuild.js` in npm v6.

`read-package-tree` evolved into `@npmcli/arborist` ([literally](https://github.com/npm/arborist/commit/b9a88513fdf415e945777c321cd4cabd62aaca81)), and both `read-installed` and `read-package-tree` were fully abandoned as we ported commands in npm/cli to be just orchestrators of calls to Arborist, which handled all the tree stuff.  (Now npm/cli's `lib/rebuild.js` just [figures out the user's request and calls `arb.rebuild()`](https://github.com/npm/cli/blob/latest/lib/rebuild.js#L43-L70).)

tl;dr - The closest parallel to `read-installed` (in the sense that it tells you what is installed) is `arborist.loadActual()`

```js
const arb = new Arborist({ path: '/some/project/path' })
const tree = await arb.loadActual()
// tree.children tells you the nodes in /some/project/path/node_modules/...
// tree.edgesOut is a map telling you what the project depends on
// most deps are Node objects, with their own .children and .edgesOut Maps
// symlink deps are represented by Link objects, which have a `link.target`, which is a Node
// tree.inventory is a Map of location (normalized relative path) to Nodes
```

Have fun!

**Reply by link89** [reply]
Thank you. That's very useful information. It would be great if you can put those information in the readme of read-installed so that others can know where to find an official replacement for it.
